vtStorAtaProtocol/Platform/Linux/ProtocolAtaPassThrough.cpp

```diff
@@ -27,164 +28,145 @@ namespace vtStor
-        m_AtaPassThrough.CurrentTaskFile[DEVICE_REGISTER_OFFSET] = CurrentTaskFile.InputRegister.Device;
-        m_AtaPassThrough.CurrentTaskFile[COMMAND_REGISTER_OFFSET] = CurrentTaskFile.InputRegister.Command;
-        m_AtaPassThrough.CurrentTaskFile[RESERVED_REGISTER_OFFSET] = CurrentTaskFile.InputRegister.Reserved;
+        m_ATAPassThrough.dxfer_direction = SG_DXFER_FROM_DEV;
+        m_Cdb16.ExtendProtocol.Protocol = static_cast<U8>(eATAProtocolFieldCdb::PIO_Data_In);
+        m_Cdb16.Param.TDir = static_cast<U8>(eTDirBit::TransFromDevice);
+    } break;
+
+    case StorageUtility::Ata::eDataAccess::WRITE_TO_DEVICE:
+    {
+        m_ATAPassThrough.dxfer_direction = SG_DXFER_TO_DEV;
+        m_Cdb16.ExtendProtocol.Protocol = static_cast<U8>(eATAProtocolFieldCdb::PIO_Data_Out);
+        m_Cdb16.Param.TDir = static_cast<U8>(eTDirBit::TransToDevice);
+    } break;
+
+    case StorageUtility::Ata::eDataAccess::NONE:
+    {
+        m_ATAPassThrough.dxfer_direction = SG_DXFER_NONE;
+        m_ATAPassThrough.dxfer_len = 0;
+        m_Cdb16.ExtendProtocol.Protocol = static_cast<U8>(eATAProtocolFieldCdb::NonData);
+        m_Cdb16.Param.CheckCondition = static_cast<U8>(eCheckCondition::CommonTerminate);
+    } break;
+
     }
 
-    eErrorCode cAtaPassThrough::IssuePassThroughDirectCommand( const DeviceHandle& Handle, U32& BytesReturned )
+    if (StorageUtility::Ata::eDataAccess::NONE != AtaCommandCharacteristic.DataAccess)
     {
-        
-        assert( INVALID_HANDLE_VALUE != Handle );
-
-        eErrorCode error;
-        error = eErrorCode::None;
-        
-        BOOL commandSuccessfulFlag;
-        DWORD bytesReturned;
-        commandSuccessfulFlag = DeviceIoControl
-            (
-            Handle,
-            IOCTL_ATA_PASS_THROUGH_DIRECT,
-            &m_AtaPassThrough,
-            m_AtaPassThrough.Length,
-            &m_AtaPassThrough,
-            m_AtaPassThrough.Length,
-            &bytesReturned,
-            NULL
-            );
-
-        // If the operation was not successful:
-        if (FALSE == commandSuccessfulFlag)
-        {
-            error = eErrorCode::Io;
-            
-            //TODO: report extended error
-            //fprintf( stderr, "\nDeviceIoControl was not successful. Error Code: %d", GetLastError() );
-        }
-
-        BytesReturned = bytesReturned;
-
-        return(error);
-    }*/
+        m_Cdb16.Param.TLength = static_cast<U8>(eTLenghValue::TLenSectorCount);
+        m_Cdb16.Param.BytBlock = static_cast<U8>(eBytBlockBit::TransSectorMode);
+    }
+
+    if (StorageUtility::Ata::eFieldFormatting::COMMAND_48_BIT == AtaCommandCharacteristic.FieldFormatting)
+    {
+        m_Cdb16.ExtendProtocol.Extend = SG_ATA_LBA48;
+    }
+
+    //! If transfer mode is DMA
+    if (StorageUtility::Ata::eTransferMode::DMA_PROTOCOL == AtaCommandCharacteristic.TransferMode)
+    {
+        m_Cdb16.ExtendProtocol.Protocol = static_cast<U8>(eATAProtocolFieldCdb::DMA);

```

Comment:
I know that. But DMA, PIO_Data_Out, PIO_Data_In or NonData is a type of Protocol. I don't want to set protocol field many times. Write a function to do it.